lib/puppet/provider/cloudfront_distribution/v2.rb

```diff
@@ -0,0 +1,241 @@
+        quantity: dist_origins.length,
+        items: dist_origins,
+      },
+      # All values below are hard-coded defaults for now.
+      default_cache_behavior: {
+        target_origin_id: dist_origins[0][:id],
+        min_ttl: 1,
+        viewer_protocol_policy: 'allow-all',
+        allowed_methods: {
+          quantity: 2,
+          items: ['GET', 'HEAD'],
+          cached_methods: {
+            quantity: 2,
+            items: ['GET', 'HEAD'],
+          },
+        },
+        forwarded_values: {
+          query_string: false,
+          cookies: {
+            forward: 'none',
+            whitelisted_names: {
+              quantity: 0,
+            },
+          },
+          headers: {
+            quantity: 0,
+          },
+          query_string_cache_keys: {
+            quantity: 0,
+          },
+        },
+        lambda_function_associations: {
+          quantity: 0,
+        },
+        trusted_signers: {
+          enabled: false,
+          quantity: 0,
+        },
+      },
+    }
+  end
+
+  def hash_to_origin(hash)
+    {
+      id: hash['id'],
+      domain_name: hash['domain_name'],
+      custom_origin_config: {
+        http_port: hash['http_port'],
+        https_port: hash['https_port'],
+        origin_protocol_policy: hash['protocol_policy'],
+        origin_ssl_protocols: {
+          quantity: hash['protocols'].length,
+          items: hash['protocols'],
+        },
+      },
+    }
+  end
+
+  def destroy
+    # Disabling the distribution returns a new etag. Otherwise, get the current one.
+    if enabled then etag = disable end
+    etag = etag ? etag : cloudfront_client.get_distribution({id: @property_hash[:id]}).etag
+
+    begin
+      cloudfront_client.delete_distribution({
+        id: @property_hash[:id],
+        if_match: etag,
+      })
+
+      @property_hash[:ensure] = :absent
+    rescue Aws::CloudFront::Errors::DistributionNotDisabled
+      Puppet.warning("The CloudFront distribution #{@property_hash[:name]} is not finished being disabled and cannot be deleted yet.")
+    end
+  end
+
+  def enabled?
+    @property_hash[:enabled] == true
+  end
+
+  def enable
+    enable_or_disable(:enable)
+  end
+
+  def disable
+    enable_or_disable(:disable)
+  end
+
+  def enable_or_disable(which)
+    to_enable = case which
+    when :enable; true
+    when :disable; false
+    else
+      fail "'#{which}' is neither :enable nor :disable"
+    end
+
+    dist_resp = cloudfront_client.get_distribution({id: @property_hash[:id]})
+    dist_resp.distribution.distribution_config.enabled = to_enable
+    dist_disabled_resp = cloudfront_client.update_distribution({
+      id: @property_hash[:id],
+      if_match: dist_resp.etag,
+      distribution_config: dist_resp.distribution.distribution_config,
+    })
+
+    @property_hash[:enabled] = to_enable
+
+    dist_disabled_resp.etag
+  end
+
+  def flush
+    Puppet.warning('Altering a CloudFront distribution is not supported yet.')
```

Comment:
This will be called after create/destroy, in addition to normal property changes. Since `create` updates `@property_hash` you may be able to set like `@property_hash[:just_created] = true` and it will otherwise get ignored by puppet but available for flush to check; I may be wrong on that. For destroy you can compare `if @property_hash[:ensure] == :absent`